sound/soc/sof/intel/hda.c

```diff
@@ -1057,8 +1116,30 @@ static int hda_sdw_machine_select(struct snd_sof_dev *sdev)
 	 */
 	if (link_mask && !pdata->machine) {
 		mach = pdata->desc->alt_machines;
-		while (mach && mach->link_mask && mach->link_mask != link_mask)
-			mach++;
+		while (mach && mach->link_mask) {
```

Comment:
you could do this as
```
for (mach = pdata->desc->alt_machines; mach && mach->link_mask; mach++) {
```
and remove both `mach++;` below. In fact, if all those arrays have an empty element `{}` in the end, you don't need to check for `mach != NULL`, provided that `mach->link_mask` *must* be `!= 0` in all valid entries.

Reply:
Good suggestion @lyakh Thanks.

Reply:
@bardliao but my worry is that we use mach == NULL as an indication that we failed to find a suitable match. Will that be affected at all?

Reply:
it's fine. mach && mach->link_mask works well, and it's the same as what we do for I2C with the HID

Reply:
@ranj063 where is that `mach == NULL` indication used? That test can only be true if the original array isn't assigned, e.g. if `pdata->desc->alt_machines == NULL` in this case. If it isn't NULL, than mach is rather unlikely to become NULL soon enough, since it's iterated per `mach++`. So it only can become NULL if it wraps around, which we hope it won't...